Add explicit list of tags to tag to docker build config and fix tag processing

jrandall opened the pull request.

jrandall committed:
Merge pull request #1 from wtsi-hgi/master

pull from upstream

jrandall committed:
add wheel to test requirements as other requirements do not install without it

jrandall committed:
add wheel to test requirements as other requirements do not install without it

jrandall committed:
add coverage to test_requirements

jrandall committed:
add coverage to test_requirements

jrandall committed:
add wheel to consul-requirements

jrandall committed:
add wheel to consul-requirements

jrandall committed:
add support for an explicit list of tags for docker build configs

jrandall committed:
add support for an explicit list of tags for docker build configs

colin-nolan commented on thriftybuilder/tests/_common.py:
```diff
@@ -181,12 +185,17 @@ def tearDown(self):
-            return False
-
+            pass
+        docker_client.close()
```
`DockerClient` should really work in a context manager to avoid this pain...

colin-nolan replied:
Thinking about it: `try... finally` is a good alternative.

jrandall replied:
I found some other issues related to this multi-tag support (wanting a mode to always push tags even when checksums match) that I've now just pushed that also touched this code. I wasn't sure how to use finally with the return values without doing basically what I already did?

colin-nolan commented on thriftybuilder/tests/_common.py:
```diff
@@ -31,7 +31,7 @@
 def create_docker_setup(
         *, commands: Iterable[str]=None, context_files: Dict[str, Optional[str]]=None,
-        image_name: str=_RANDOM_NAME, from_image_name: str=EXAMPLE_FROM_IMAGE_NAME) \
+        image_name: str=_RANDOM_NAME, tags: List[str]=[], from_image_name: str=EXAMPLE_FROM_IMAGE_NAME) \
```
Note: mutable default arguments are to be avoided as they are evaluated only when the function is defined.
```
>>> def append_to_list(element, list=[]):
...     list.append(element)
...     return list
... 
>>> append_to_list(1)
[1]
>>> append_to_list(2)
[1, 2]
```

colin-nolan replied:
Changed in https://github.com/wtsi-hgi/thrifty-builder/commit/6235b3f9a84fe02c2488e5b7885e924ba4fc65c9.

jrandall replied:
incorporated 6235b3f9a84fe02c2488e5b7885e924ba4fc65c9

Commit:
Allows `DockerUploader` to be used in a context manager.

Commit:
Minor changes.

Commit:
Minor changes.

colin-nolan commented on thriftybuilder/build_configurations.py:
```diff
@@ -160,6 +162,14 @@ def __init__(self, image_name: str, dockerfile_location: str, context: str=None)
+            name_tag = self._identifier.split(DockerBuildConfiguration._NAME_TAG_SEPARATOR)
+            self._identifier = name_tag[0]
+            tag = name_tag[1]
+            if len(tag) > 0:
```
This is going to be a truism so long as `image_name` has not been set to `myimage:` (which should be rejected as it's not a valid name).

colin-nolan commented on thriftybuilder/build_configurations.py:
```diff
@@ -160,6 +162,14 @@ def __init__(self, image_name: str, dockerfile_location: str, context: str=None)
+            self._identifier = name_tag[0]
+            tag = name_tag[1]
+            if len(tag) > 0:
+                self._tags.insert(0, tag)
```
The computer scientist in my wants this to be a set to avoid duplicate tags. As far as I can tell, there is nothing special about the first tag in `self._tags`

jrandall replied:
agree we should avoid duplicate tags, i'll add that now (though perhaps more horribly than you'd like)

jrandall replied:
fixed in eb3bcef0776cc78f0cd32b183af6c48f24781344

colin-nolan replied:
Please see my PR!!
https://github.com/jrandall/thrifty-builder/pull/2

jrandall replied:
seen and merged

Commit:
Tests uploading multiple tags.

Commit:
Change to storing tags in set.

Commit:
Adds a few extra tests.

jrandall committed:
add always_upload (default False) to docker build configuration to force uploading of new tags on preexisting checksums

Commit:
Bug fix.

jrandall committed:
Merge branch 'master' into docker-config-tag-list

jrandall committed:
avoids duplicate tags when using legacy name-based tagging

avoids having identical tags when e.g.
```
name: "foo/bar:latest"
tags: ["latest"]
```

colin-nolan commented:
See https://github.com/jrandall/thrifty-builder/pull/2 (unfortunately I couldn't add commits to this PR so I'm afraid there's a few conflicts with your latest commits).

jrandall committed:
Merge branch 'docker-config-tag-list' into docker-config-tag-list

jrandall committed:
Merge pull request #2 from wtsi-hgi/docker-config-tag-list

Store tags as set and some minor changes

jrandall committed:
fix indentation error in merge of with

jrandall commented:
there were some merging issues that reverted some earlier work. fixing them now.

jrandall commented:
oops, did not mean to merge this to wtsi-hgi master, only to my own - sorry @colin-nolan
